pkg/k8s/watchers/cilium_node.go

```diff
@@ -5,167 +5,174 @@ package watchers
-							if oldCN.DeepEqual(ciliumNode) &&
-								comparator.MapStringEquals(oldCN.ObjectMeta.Labels, ciliumNode.ObjectMeta.Labels) &&
-								comparator.MapStringEquals(oldCN.ObjectMeta.Annotations, ciliumNode.ObjectMeta.Annotations) {
-								equal = true
-								if !isLocal {
-									// For remote nodes, we return early here to avoid unnecessary update events if
-									// nothing in the spec or status has changed. But for local nodes, we want to
-									// propagate the new resource version (not compared in DeepEqual) such that any
-									// CiliumNodeChain subscribers are able to perform updates to the local CiliumNode
-									// object using the most recent resource version.
-									return
-								}
-							}
-							n := nodeTypes.ParseCiliumNode(ciliumNode)
-							if isLocal {
-								return
-							}
-							k.nodeDiscoverManager.NodeUpdated(n)
-							k.K8sEventProcessed(metricCiliumNode, resources.MetricUpdate, true)
-						}
-					}
-				},
-				DeleteFunc: func(obj interface{}) {
-					var equal bool
-					defer func() { k.K8sEventReceived(apiGroup, metricCiliumNode, resources.MetricDelete, true, equal) }()
-					ciliumNode := k8s.CastInformerEvent[cilium_v2.CiliumNode](obj)
-					if ciliumNode == nil {
-						return
-					}
-					n := nodeTypes.ParseCiliumNode(ciliumNode)
-					k.nodeDiscoverManager.NodeDeleted(n)
-				},
-			},
+		var synced atomic.Bool
+		stop := make(chan struct{})
+
+		k.blockWaitGroupToSyncResources(
+			stop,
 			nil,
+			func() bool { return synced.Load() },
+			apiGroup,
 		)
-		isConnected := make(chan struct{})
-		// once isConnected is closed, it will stop waiting on caches to be
-		// synchronized.
-		k.blockWaitGroupToSyncResources(isConnected, swgNodes, ciliumNodeInformer.HasSynced, apiGroup)
-
-		k.ciliumNodeStoreMU.Lock()
-		k.ciliumNodeStore = ciliumNodeStore
-		k.ciliumNodeStoreMU.Unlock()
-
-		once.Do(func() {
-			// Signalize that we have put node controller in the wait group
-			// to sync resources.
-			asyncControllers.Done()
-		})
 		k.k8sAPIGroups.AddAPI(apiGroup)
-		go ciliumNodeInformer.Run(isConnected)
 
-		<-kvstore.Connected()
-		log.Info("Connected to key-value store, stopping CiliumNode watcher")
+		// Signalize that we have put node controller in the wait group to sync resources.
+		once.Do(asyncControllers.Done)
+
+		// derive another context to signal Events() and Store() in case of kvstore connection
+		subCtx, cancel := context.WithCancel(ctx)
+
+		var wg sync.WaitGroup
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer close(stop)
+
+			events := k.resources.CiliumNode.Events(subCtx)
+			cache := make(map[resource.Key]*cilium_v2.CiliumNode)
+			for event := range events {
+				var err error
+				switch event.Kind {
+				case resource.Sync:
+					synced.Store(true)
+				case resource.Upsert:
+					var needUpdate bool
+					oldObj, ok := cache[event.Key]
```

Comment:
Is this "insert" stuff really needed? `onCiliumNodeUpdate` calls `onCiliumNodeInsert`. Why don't we just have update?

Reply:
Without the cache we couldn't filter out "update" events for which `oldNode.DeepEqual(newNode) && maps.Equal(...)` is true, thus calling `types.ParseCiliumNode` and `k.nodeDiscoverManager.NodeUpdated(n)` more times than needed.
TBH I'm not sure about the performance impact of that, but being a refactoring I chose to keep the old behavior.